### PR TITLE
ISPN-4075 State transfer should preserve the timestamps of entries 

### PR DESCRIPTION
**Pull request by danberindei**
https://issues.jboss.org/browse/ISPN-4075

The fix is a bit crude because I didn't want to add the timestamps in regular PutKeyValueCommands, but it does the job.

### CODE REVIEW

File: core/src/main/java/org/infinispan/commands/write/PutForStateTransferCommand.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
+package org.infinispan.commands.write;
+
+import org.infinispan.commands.CommandInvocationId;
+import org.infinispan.metadata.Metadata;
+
+public class PutForStateTransferCommand extends PutKeyValueCommand {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/commands/write/PutForStateTransferCommand.java]: I am thinking we should have this class should implement `LocalCommand` so we can easily see which commands aren't to be replicated.

**Comment by danberindei** [comment]
I realized I had forgotten about scattered caches, and then I noticed they already use a different approach: a `PutMapCommand(PUT_FOR_STATE_TRANSFER)` with `InternalCacheValue` instead of regular values. So I changed my code to use `InternalCacheEntry` values instead of adding a new command type (ICE to avoid allocating an ICV instance in `ICE.toInternalCacheValue()`).

File: core/src/main/java/org/infinispan/container/impl/InternalDataContainer.java [hunk 2]
```diff
@@ -47,12 +47,19 @@
...
     * @param k key under which to store entry
     * @param v value to store
     * @param metadata metadata of the entry
+    * @param createTimestamp creation timestamp, or 0 to use the current time
+    * @param lastUseTimestamp last use timestamp, or 0 to use the current time
+    *
+    * @since 10.0
     */
-   void put(int segment, K k, V v, Metadata metadata);
+   void put(int segment, K k, V v, Metadata metadata, long createTimestamp, long lastUseTimestamp);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/container/impl/InternalDataContainer.java]: I am glad we added this internal interface :)

File: core/src/main/java/org/infinispan/container/offheap/SegmentedBoundedOffHeapDataContainer.java [hunk 3]
```diff
@@ -97,8 +97,8 @@ public void put(WrappedBytes key, WrappedBytes value, Metadata metadata) {
    }
 
    @Override
-   public void put(int segment, WrappedBytes wrappedBytes, WrappedBytes wrappedBytes2, Metadata metadata) {
-      super.put(segment, wrappedBytes, wrappedBytes2, metadata);
+   public void put(int segment, WrappedBytes wrappedBytes, WrappedBytes wrappedBytes2, Metadata metadata, long createTimestamp, long lastUseTimestamp) {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/container/offheap/SegmentedBoundedOffHeapDataContainer.java]:
```suggestion
   public void put(int segment, WrappedBytes key, WrappedBytes value, Metadata metadata, long createTimestamp, long lastUseTimestamp) {
```
Oops. Didn't realize these were messed up before.

File: core/src/main/java/org/infinispan/container/offheap/SegmentedBoundedOffHeapDataContainer.java [hunk 4]
```diff
@@ -97,8 +97,8 @@ public void put(WrappedBytes key, WrappedBytes value, Metadata metadata) {
...
+      super.put(segment, wrappedBytes, wrappedBytes2, metadata, createTimestamp, lastUseTimestamp);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/container/offheap/SegmentedBoundedOffHeapDataContainer.java]:
```suggestion
      super.put(segment, key, value, metadata, createTimestamp, lastUseTimestamp);
```

File: core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java [hunk 5]
```diff
@@ -641,7 +641,8 @@ private void doApplyState(Address sender, int segmentId, Collection<InternalCach
                ctx = icf.createSingleKeyNonTxInvocationContext();
             }
 
-            PutKeyValueCommand put = commandsFactory.buildPutKeyValueCommand(e.getKey(), e.getValue(), segmentId,
+            // CallInterceptor knows PUT_FOR_STATE_TRANSFER values are actually InternalCacheEntries
+            PutKeyValueCommand put = commandsFactory.buildPutKeyValueCommand(e.getKey(), e, segmentId,
```
Comment (wburns) [inline core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java]: The only thing I wonder is, we could be duplicating the metadata we send across right? Maybe we should pass null for metadata?

File: core/src/test/java/org/infinispan/statetransfer/StateTransferTimestampsTest.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+import org.infinispan.test.MultipleCacheManagersTest;
+import org.infinispan.test.TestingUtil;
+import org.infinispan.util.ControlledTimeService;
+import org.testng.annotations.Test;
+
+/**
+ * Tests concurrent startup of replicated and distributed caches
+ *
+ * @author Dan Berindei
+ * @since 7.2
```
Comment (wburns) [inline core/src/test/java/org/infinispan/statetransfer/StateTransferTimestampsTest.java]: Wrong version :)

File: core/src/test/java/org/infinispan/statetransfer/StateTransferTimestampsTest.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+ */
+@Test(testName = "statetransfer.StateTransferTimestampsTest", groups = "functional")
+public class StateTransferTimestampsTest extends MultipleCacheManagersTest {
+   public static final String CACHE_NAME = "cache";
+   private ControlledTimeService timeService;
+
+   @Override
+   public Object[] factory() {
+      return new Object[]{
+         new StateTransferTimestampsTest().cacheMode(CacheMode.REPL_SYNC),
```
Comment (wburns) [inline core/src/test/java/org/infinispan/statetransfer/StateTransferTimestampsTest.java]: No DIST_SYNC ?

File: core/src/test/java/org/infinispan/statetransfer/StateTransferTimestampsTest.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+      };
+   }
+
+   @Override
+   protected void createCacheManagers() throws Throwable {
+      createCluster(new ConfigurationBuilder(), 2);
+
+      timeService = new ControlledTimeService();
+      ConfigurationBuilder replConfig = new ConfigurationBuilder();
+      replConfig.clustering().cacheMode(CacheMode.REPL_SYNC).hash().numSegments(4);
```
Comment (wburns) [inline core/src/test/java/org/infinispan/statetransfer/StateTransferTimestampsTest.java]: Seems we aren't testing Scattered then?

File: core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java [hunk 9]
```diff
@@ -73,6 +73,7 @@
 import org.infinispan.container.entries.CacheEntry;
 import org.infinispan.container.entries.ExpiryHelper;
 import org.infinispan.container.entries.InternalCacheEntry;
+import org.infinispan.container.entries.InternalCacheValue;
```
Comment (wburns) [inline core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java]: Pesky checkstyle.

File: core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java (hunk 5 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java]: These put commands are never sent over the network.

File: core/src/test/java/org/infinispan/statetransfer/StateTransferTimestampsTest.java (hunk 6 above)
Comment (danberindei) [reply core/src/test/java/org/infinispan/statetransfer/StateTransferTimestampsTest.java]: Oops!

File: core/src/test/java/org/infinispan/statetransfer/StateTransferTimestampsTest.java (hunk 7 above)
Comment (danberindei) [reply core/src/test/java/org/infinispan/statetransfer/StateTransferTimestampsTest.java]: The state transfer process is the same for all clustered caches except scattered.

File: core/src/test/java/org/infinispan/statetransfer/StateTransferTimestampsTest.java (hunk 8 above)
Comment (danberindei) [reply core/src/test/java/org/infinispan/statetransfer/StateTransferTimestampsTest.java]: I thought I was just lucky that it passed the first time I ran the test :)

**Comment by danberindei** [comment]
Ok, fixing scattered mode took longer than I thought, but it works now.

**Comment by wburns** [comment]
Changes look good to me, however CI seems to have puked https://ci.infinispan.org/job/Infinispan/job/PR-6525/9/consoleFull#-897946742ea5a4324-14c7-4883-b210-e33383209aa7
I haven't looked recently if this is related or not.

**Comment by wburns** [comment]
👍

**Comment by wburns** [comment]
Integrated into master, thanks @danberindei !

**Pull request merged**